core/toolbox/toolbox.js

```diff
@@ -0,0 +1,463 @@
+        'oneBasedIndex': workspace.options.oneBasedIndex,
+        'horizontalLayout': workspace.horizontalLayout,
+        'renderer': workspace.options.renderer,
+        'rendererOverrides': workspace.options.rendererOverrides
+      }));
+  workspaceOptions.toolboxPosition = workspace.options.toolboxPosition;
+
+  if (workspace.horizontalLayout) {
+    if (!Blockly.HorizontalFlyout) {
+      throw Error('Missing require for Blockly.HorizontalFlyout');
+    }
+    flyout = new Blockly.HorizontalFlyout(workspaceOptions);
+  } else {
+    if (!Blockly.VerticalFlyout) {
+      throw Error('Missing require for Blockly.VerticalFlyout');
+    }
+    flyout = new Blockly.VerticalFlyout(workspaceOptions);
+  }
+  if (!flyout) {
+    throw Error('One of Blockly.VerticalFlyout or Blockly.Horizontal must be' +
+        'required.');
+  }
+  return flyout;
+};
+
+/**
+ * Adds all the toolbox items to the toolbox.
+ * @param {!Array<!Blockly.utils.toolbox.ToolboxItemDef>} toolboxDef Array
+ *     holding objects containing information on the contents of the toolbox.
+ * @protected
+ */
+Blockly.Toolbox.prototype.addContents_ = function(toolboxDef) {
+  for (var i = 0, childIn; (childIn = toolboxDef[i]); i++) {
+    // TODO: Add classes to registry so we can avoid this switch statement.
+    switch (childIn['kind'].toUpperCase()) {
+      case 'CATEGORY':
+        childIn = /** @type {Blockly.utils.toolbox.Category} */ (childIn);
+        var category = new Blockly.ToolboxCategory(childIn, this);
+        this.addToolboxItem_(category);
+        var categoryDom = category.createDom();
+        if (categoryDom) {
+          this.contentsDiv_.appendChild(categoryDom);
+        }
+        break;
+      case 'SEP':
+        childIn = /** @type {Blockly.utils.toolbox.Separator} */ (childIn);
+        var separator = new Blockly.ToolboxSeparator(childIn, this);
+        this.addToolboxItem_(separator);
+        var separatorDom = separator.createDom();
+        if (separatorDom) {
+          this.contentsDiv_.appendChild(separatorDom);
+        }
+        break;
+      default:
+        // TODO: Handle someone adding a custom component.
+    }
+  }
+};
+
+/**
+ * Fills the toolbox with new toolbox items and removes any old contents.
+ * @param {!Array<!Blockly.utils.toolbox.ToolboxItemDef>} toolboxDef Array
+ *     holding objects containing information on the contents of the toolbox.
+ * @package
+ */
+Blockly.Toolbox.prototype.render = function(toolboxDef) {
+  this.toolboxDef_ = toolboxDef;
+  // TODO: Future improvement to compare the new toolboxDef with the old and
+  //  only re render what has changed.
+  for (var i = 0; i < this.contents_.length; i++) {
+    var child = this.contents_[i];
+    child.getDiv().remove();
+  }
+  this.contents_ = [];
+  this.contentIds_ = {};
+  this.addContents_(toolboxDef);
+  this.position();
+};
+
+/**
+ * Adds an item to the toolbox.
+ * @param {!Blockly.ToolboxItem} toolboxItem The item in the toolbox.
+ * @private
```

Comment:
Why is this private, whereas addContents is protected.

Reply:
mm I don't think my reasons are relevant anymore. Fixed.